### DISCUSSION TITLE
Getting Started

### DISCUSSION THREAD

**Discussion by oscarrutt** [Homebrew/discussions#5747] [Getting started] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.4.5
ORIGIN: https://github.com/Homebrew/brew
HEAD: 254bf3fe9d8fa2e1b2fb55dbcf535b2d870180c4
Last commit: 9 days ago
Core tap JSON: 19 Nov 03:32 UTC
Core cask tap JSON: 19 Nov 03:32 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.JM15ADJCeC/org.xquartz:0
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 3.3.6 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.6/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 16.0.0 build 1600
Git: 2.39.5 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 14.7.1-arm64
CLT: 16.1.0.0.1.1729049160
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/libcsirocsa.0.0.1.dylib
  /usr/local/lib/libg2c.0.0.0.dylib
  /usr/local/lib/libgcc_s.1.0.dylib
  /usr/local/lib/libgmp.3.4.2.dylib
  /usr/local/lib/libgmp.3.dylib
  /usr/local/lib/libgmp.dylib
  /usr/local/lib/libmpfr.1.1.1.dylib
  /usr/local/lib/libmpfr.1.dylib
  /usr/local/lib/libmpfr.dylib
  /usr/local/lib/libneon.25.0.5.dylib
  /usr/local/lib/libsvn_client-1.0.0.0.dylib
  /usr/local/lib/libsvn_delta-1.0.0.0.dylib
  /usr/local/lib/libsvn_diff-1.0.0.0.dylib
  /usr/local/lib/libsvn_fs-1.0.0.0.dylib
  /usr/local/lib/libsvn_fs_fs-1.0.0.0.dylib
  /usr/local/lib/libsvn_ra-1.0.0.0.dylib
  /usr/local/lib/libsvn_ra_dav-1.0.0.0.dylib
  /usr/local/lib/libsvn_ra_local-1.0.0.0.dylib
  /usr/local/lib/libsvn_ra_svn-1.0.0.0.dylib
  /usr/local/lib/libsvn_repos-1.0.0.0.dylib
  /usr/local/lib/libsvn_subr-1.0.0.0.dylib
  /usr/local/lib/libsvn_wc-1.0.0.0.dylib
  /usr/local/lib/libsvnjavahl-1.0.0.0.dylib

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae and may need to be deleted.

Unexpected header files:
  /usr/local/include/faust/au/AUUI.h
  /usr/local/include/faust/audio/alsa-dsp.h
  /usr/local/include/faust/audio/audio.h
  /usr/local/include/faust/audio/coreaudio-dsp.h
  /usr/local/include/faust/audio/coreaudio-ios-dsp.h
  /usr/local/include/faust/audio/dsp-adapter.h
  /usr/local/include/faust/audio/dsp.h
  /usr/local/include/faust/audio/jack-dsp.h
  /usr/local/include/faust/audio/netjack-dsp.h
  /usr/local/include/faust/audio/oscdsp.h
  /usr/local/include/faust/audio/portaudio-dsp.h
  /usr/local/include/faust/gui/ControlUI.h
  /usr/local/include/faust/gui/FUI.h
  /usr/local/include/faust/gui/GUI.h
  /usr/local/include/faust/gui/HTTPDControler.h
  /usr/local/include/faust/gui/JSONUI.h
  /usr/local/include/faust/gui/MapUI.h
  /usr/local/include/faust/gui/OSCControler.h
  /usr/local/include/faust/gui/OSCUI.h
  /usr/local/include/faust/gui/PathUI.h
  /usr/local/include/faust/gui/PrintUI.h
  /usr/local/include/faust/gui/UI.h
  /usr/local/include/faust/gui/console.h
  /usr/local/include/faust/gui/faustgtk.h
  /usr/local/include/faust/gui/faustqt.h
  /usr/local/include/faust/gui/httpdUI.h
  /usr/local/include/faust/gui/jsonfaustui.h
  /usr/local/include/faust/gui/meta.h
  /usr/local/include/faust/misc.h
  /usr/local/include/faust/osc/FaustFactory.h
  /usr/local/include/faust/osc/FaustNode.h
  /usr/local/include/faust/osc/Message.h
  /usr/local/include/faust/osc/MessageDriven.h
  /usr/local/include/faust/osc/MessageProcessor.h
  /usr/local/include/faust/osc/smartpointer.h
  /usr/local/include/faust/vst/faust.h
  /usr/local/include/faust/vst/voice.h
  /usr/local/include/faust/vst/vstui.h
  /usr/local/include/jack/control.h
  /usr/local/include/jack/intclient.h
  /usr/local/include/jack/jack.h
  /usr/local/include/jack/jslist.h
  /usr/local/include/jack/metadata.h
  /usr/local/include/jack/midiport.h
  /usr/local/include/jack/net.h
  /usr/local/include/jack/ringbuffer.h
  /usr/local/include/jack/session.h
  /usr/local/include/jack/statistics.h
  /usr/local/include/jack/systemdeps.h
  /usr/local/include/jack/thread.h
  /usr/local/include/jack/transport.h
  /usr/local/include/jack/types.h
  /usr/local/include/jack/uuid.h
  /usr/local/include/jack/weakjack.h
  /usr/local/include/jack/weakmacros.h
  /usr/local/include/neon/ne_207.h
  /usr/local/include/neon/ne_acl.h
  /usr/local/include/neon/ne_alloc.h
  /usr/local/include/neon/ne_auth.h
  /usr/local/include/neon/ne_basic.h
  /usr/local/include/neon/ne_compress.h
  /usr/local/include/neon/ne_dates.h
  /usr/local/include/neon/ne_defs.h
  /usr/local/include/neon/ne_i18n.h
  /usr/local/include/neon/ne_locks.h
  /usr/local/include/neon/ne_md5.h
  /usr/local/include/neon/ne_props.h
  /usr/local/include/neon/ne_redirect.h
  /usr/local/include/neon/ne_request.h
  /usr/local/include/neon/ne_session.h
  /usr/local/include/neon/ne_socket.h
  /usr/local/include/neon/ne_ssl.h
  /usr/local/include/neon/ne_string.h
  /usr/local/include/neon/ne_uri.h
  /usr/local/include/neon/ne_utils.h
  /usr/local/include/neon/ne_xml.h
  /usr/local/include/neon/ne_xmlreq.h
  /usr/local/include/plplot/disptab.h
  /usr/local/include/plplot/drivers.h
  /usr/local/include/plplot/gcw.h
  /usr/local/include/plplot/pdf.h
  /usr/local/include/plplot/plConfig.h
  /usr/local/include/plplot/plDevs.h
  /usr/local/include/plplot/pldebug.h
  /usr/local/include/plplot/pldll.h
  /usr/local/include/plplot/plevent.h
  /usr/local/include/plplot/plplot.h
  /usr/local/include/plplot/plplotP.h
  /usr/local/include/plplot/plplotcanvas.h
  /usr/local/include/plplot/plstrm.h
  /usr/local/include/plplot/plxwd.h
  /usr/local/include/readline/chardefs.h
  /usr/local/include/readline/history.h
  /usr/local/include/readline/keymaps.h
  /usr/local/include/readline/readline.h
  /usr/local/include/readline/rlconf.h
  /usr/local/include/readline/rlstdc.h
  /usr/local/include/readline/rltypedefs.h
  /usr/local/include/readline/tilde.h
  /usr/local/include/subversion-1/mod_dav_svn.h
  /usr/local/include/subversion-1/svn_auth.h
  /usr/local/include/subversion-1/svn_base64.h
  /usr/local/include/subversion-1/svn_client.h
  /usr/local/include/subversion-1/svn_cmdline.h
  /usr/local/include/subversion-1/svn_config.h
  /usr/local/include/subversion-1/svn_ctype.h
  /usr/local/include/subversion-1/svn_dav.h
  /usr/local/include/subversion-1/svn_delta.h
  /usr/local/include/subversion-1/svn_diff.h
  /usr/local/include/subversion-1/svn_error.h
  /usr/local/include/subversion-1/svn_error_codes.h
  /usr/local/include/subversion-1/svn_fs.h
  /usr/local/include/subversion-1/svn_hash.h
  /usr/local/include/subversion-1/svn_io.h
  /usr/local/include/subversion-1/svn_md5.h
  /usr/local/include/subversion-1/svn_nls.h
  /usr/local/include/subversion-1/svn_opt.h
  /usr/local/include/subversion-1/svn_path.h
  /usr/local/include/subversion-1/svn_pools.h
  /usr/local/include/subversion-1/svn_props.h
  /usr/local/include/subversion-1/svn_quoprint.h
  /usr/local/include/subversion-1/svn_ra.h
  /usr/local/include/subversion-1/svn_ra_svn.h
  /usr/local/include/subversion-1/svn_repos.h
  /usr/local/include/subversion-1/svn_sorts.h
  /usr/local/include/subversion-1/svn_string.h
  /usr/local/include/subversion-1/svn_subst.h
  /usr/local/include/subversion-1/svn_time.h
  /usr/local/include/subversion-1/svn_types.h
  /usr/local/include/subversion-1/svn_utf.h
  /usr/local/include/subversion-1/svn_version.h
  /usr/local/include/subversion-1/svn_wc.h
  /usr/local/include/subversion-1/svn_xml.h

Warning: Unbrewed '.la' files were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae and may need to be deleted.

Unexpected '.la' files:
  /usr/local/lib/libg2c.la
  /usr/local/lib/libgmp.la
  /usr/local/lib/libmpfr.la
  /usr/local/lib/libneon.la
  /usr/local/lib/libsvn_client-1.la
  /usr/local/lib/libsvn_delta-1.la
  /usr/local/lib/libsvn_diff-1.la
  /usr/local/lib/libsvn_fs-1.la
  /usr/local/lib/libsvn_fs_fs-1.la
  /usr/local/lib/libsvn_ra-1.la
  /usr/local/lib/libsvn_ra_dav-1.la
  /usr/local/lib/libsvn_ra_local-1.la
  /usr/local/lib/libsvn_ra_svn-1.la
  /usr/local/lib/libsvn_repos-1.la
  /usr/local/lib/libsvn_subr-1.la
  /usr/local/lib/libsvn_wc-1.la
  /usr/local/lib/libsvnjavahl-1.la

Warning: Unbrewed '.pc' files were found in /usr/local/lib/pkgconfig.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae and may need to be deleted.

Unexpected '.pc' files:
  /usr/local/lib/pkgconfig/jack.pc
  /usr/local/lib/pkgconfig/neon.pc

Warning: Unbrewed static libraries were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae and may need to be deleted.

Unexpected static libraries:
  /usr/local/lib/libfrtbegin.a
  /usr/local/lib/libg2c.a
  /usr/local/lib/libgmp.a
  /usr/local/lib/libhistory.a
  /usr/local/lib/libmpfr.a
  /usr/local/lib/libneon.a
  /usr/local/lib/libreadline.a
  /usr/local/lib/libsvn_client-1.a
  /usr/local/lib/libsvn_delta-1.a
  /usr/local/lib/libsvn_diff-1.a
  /usr/local/lib/libsvn_fs-1.a
  /usr/local/lib/libsvn_fs_fs-1.a
  /usr/local/lib/libsvn_ra-1.a
  /usr/local/lib/libsvn_ra_dav-1.a
  /usr/local/lib/libsvn_ra_local-1.a
  /usr/local/lib/libsvn_ra_svn-1.a
  /usr/local/lib/libsvn_repos-1.a
  /usr/local/lib/libsvn_subr-1.a
  /usr/local/lib/libsvn_wc-1.a
  /usr/local/lib/libsvnjavahl-1.a
```

### Description of issue

I have spent several hours installing Homebrew on macOS Sonoma and trying to get things to work the way I expect, but without success.

I understand that with Homebrew and Python > 3.12 I should use a virtual Python environment to manage external libraries such as the Python binding to Plotly.

I want to:

1. Run Python or a Python script from the command line and import both NumPy and Plotly.

2. Call Python to run a Python script from a C program using the `system` command. (OK, it’s really an Ada program that is bound to`system` but that’s not important.) The Python script imports both NumPy and Plotly.

I understand that I should install NumPy using brew and use pip or pip3 to install Plotly, an externally-managed library, into the virtual environment.

Here is my latest attempt:

* Install Homebrew
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)”
— Add this line to my ~/.profile:
eval "$(/opt/homebrew/bin/brew shellenv)"

* Install Homebrew Packages (e.g., NumPy)
brew install numpy

* Install a Virtual Environment at /opt/Virtual_Environment
cd /opt
python3 -m venv Virtual_Environment

* Install Plotly in the Virtual Environment
  * Activate the virtual environment:
source /opt/Virtual_Environment/bin/activate
  * Install Plotly using pip
pip install plotly

* Make this C test program:
#include <stdlib.h>
int main() {
    // Call the Python script
    system("/opt/Virtual_Environment/bin/python3 test_plotly.py");
    return 0;
}

and this Python script:
import plotly
import numpy
print("Success importing plotly and numpy”)

* This causes the numpy library to not be found:
ModuleNotFoundError: No module named ‘numpy'

* Same result if I first use
source /opt/Virtual_Environment/bin/activate
from either the command line or from my .profile file.

* The Homebrew Python is being used:
>>> import sys
>>> print(sys.executable)
/opt/Virtual_Environment/bin/python3
which points to
/opt/homebrew/Cellar/python@3.13/3.13.0_1/Frameworks/Python.framework/Versions/3.13/bin/python3.13

* Check PATH
$ echo $PATH
returns
/opt/homebrew/bin:/opt/homebrew/sbin: at the beginning plus a bunch of other stuff.

*** How do I get this test to work with numpy?

Thanks
Jerry

**Comment by p-linnane** [comment]
It's probably easier to use `pyenv` or `pipx` rather than our Python.

**Comment by gromgit** [comment, marked as answer]
https://docs.brew.sh/Homebrew-and-Python#pep-668-and-virtual-environments seems relevant. 

> brew install numpy

You probably want to install with pip, as recommended in the above page.

**Reply by oscarrutt** [reply]
@gromgit Thanks for that guidance. I read read that passage multiple times and it just didn't sink in for some reason. I uninstalled the brew numpy and installed it in the virtual environment. If I recall, the correct(ing) steps were:
```
$brew uninstall numpy
$/path/to/virtual_environment/bin/activate
(virtual_environment) $pip install numpy
```
So this installs numpy in the site-packages of the virtual environment rather than in the brew Python site-packages—the desired separation.

I'll consider this closed, but I have one remaining question. After running `brew uninstall numpy`, why is there still this: `/opt/homebrew/lib/python3.13/site-packages/numpy` ?

Thanks

**Reply by gromgit** [reply]
If it still exists, it's likely not empty. What's in that directory? Were there any warnings or errors with `brew uninstall bumpy`?

**Reply by oscarrutt** [reply]
`uninstall numpy` reports  this error:
`$ brew uninstall numpy
Error: No such keg: /opt/homebrew/Cellar/numpy`
Indeed, there is no numpy in Cellar.

After running `uninstall`, `/opt/homebrew/lib/python3.13/site-packages/numpy` contains 1.9 MB and has this as its top level:

/opt/homebrew/lib/python3.13/site-packages/numpy
/opt/homebrew/lib/python3.13/site-packages/numpy/__pycache__
/opt/homebrew/lib/python3.13/site-packages/numpy/_core
/opt/homebrew/lib/python3.13/site-packages/numpy/_typing
/opt/homebrew/lib/python3.13/site-packages/numpy/_utils
/opt/homebrew/lib/python3.13/site-packages/numpy/lib
/opt/homebrew/lib/python3.13/site-packages/numpy/linalg
/opt/homebrew/lib/python3.13/site-packages/numpy/matrixlib

Maybe numpy is a dependent of another brew installation.

**Reply by gromgit** [reply]
> `/opt/homebrew/lib/python3.13/site-packages/numpy` contains 1.9 MB 

Is everything under this hierarchy symlinks to Homebrew `Cellar` directories? If not (and the size you quoted suggests it isn't), then you probably ran `pip install numpy` outside your venv, which would've installed Numpy right there.

In summary:
* `brew install numpy` installs files in `/opt/homebrew/Cellar/numpy/<ver>/lib/pythonXXX/site-packages` and symlinks then into `/opt/homebrew/lib/pythonXXX/site-packages`
* `pip install numpy` at the global level installs files directly into `/opt/homebrew/lib/pythonXXX/site-packages`
* `pip install numpy` in a venv installs files in the `site-packages` directory of that venv.

**Reply by oscarrutt** [reply]
@gromgit You are right--no symlinks in this directory, only .pyc. `/opt/homebrew/lib/python3.13` is also not a symlink but a regular directory. (Also `/opt/homebrew/lib/python3.12`) Can I just trash these directories?
`$ pip3 show numpy` reports `WARNING: Package(s) not found: numpy`.

Also, `pip3 uninstall numpy` results only in the message starting with `This environment is externally managed' etc. The directory `/opt/homebrew/lib/python3.13` remains, along with its numpy.

Sorry for drawing this out.

**Reply by gromgit** [reply]
Try `pip3 uninstall --break-system-packages numpy`.

**Reply by oscarrutt** [reply]
~$ pip3 uninstall --break-system-packages numpy
WARNING: Skipping numpy as it is not installed.`
(sigh)

**Reply by gromgit** [reply]
Hmmm. With the major caveat that I don't do Python (yet), I guess you could `rm -fr /opt/homebrew/lib/python3.13/site-packages/numpy`, since that directory is no longer recognised by both `brew` and Homebrew Python.

**Comment by oscarrutt** [comment]
I just now deleted /opt/homebrew/lib/python3.13 and /opt/homebrew/lib/python3.12. I'll see if anything breaks

(I don't really do Python either but I found the Plotly plotting package which I really like and the best way to use it from an external compiled language is through Python. Julia works too except for its awful time-to-first-plot problem.)
